vault/identity_store_oidc_provider.go

```diff
@@ -1504,20 +1588,473 @@ func authResponse(code, state, errorCode, errorDescription string) (*logical.Res
+	authCodeEntryRaw, ok, err := i.oidcAuthCodeCache.Get(ns, code)
+	defer i.oidcAuthCodeCache.Delete(ns, code)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+	if !ok {
+		return tokenResponse(nil, ErrTokenInvalidGrant, "authorization grant is invalid or expired")
+	}
+	authCodeEntry, ok := authCodeEntryRaw.(*authCodeCacheEntry)
+	if !ok {
+		return tokenResponse(nil, ErrTokenServerError, "failed type assertion for authorization code entry")
+	}
+
+	// Ensure the authorization code was issued to the authenticated client
+	if authCodeEntry.clientID != clientID {
+		return tokenResponse(nil, ErrTokenInvalidGrant, "authorization grant is invalid or expired")
+	}
+
+	// Ensure that the redirect_uri parameter value is identical to the redirect_uri
+	// parameter value that was included in the initial authorization request.
+	redirectURI := d.Get("redirect_uri").(string)
+	if redirectURI == "" {
+		return tokenResponse(nil, ErrTokenInvalidRequest, "redirect_uri parameter is required")
+	}
+	if authCodeEntry.redirectURI != redirectURI {
+		return tokenResponse(nil, ErrTokenInvalidGrant, "redirect_uri does not match the redirect_uri used in the authorization request")
+	}
+
+	// Get the entity associated with the initial authorization request
+	entity, err := i.MemDBEntityByID(authCodeEntry.entityID, true)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+	if entity == nil {
+		return tokenResponse(nil, ErrTokenInvalidRequest, "identity entity associated with request not found")
+	}
+
+	// Validate that the entity is a member of the client's assignments
+	isMember, err := i.entityHasAssignment(ctx, req.Storage, entity.GetID(), client.Assignments)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+	if !isMember {
+		return tokenResponse(nil, ErrTokenInvalidRequest, "identity entity not authorized by client assignment")
+	}
+
+	// The access token is a Vault batch token with a policy that only
+	// provides access to the issuing provider's userinfo endpoint.
+	accessTokenIssuedAt := time.Now()
+	accessTokenExpiry := accessTokenIssuedAt.Add(client.AccessTokenTTL)
+	accessToken := &logical.TokenEntry{
+		Type:         logical.TokenTypeBatch,
+		NamespaceID:  ns.ID,
+		Path:         req.Path,
+		TTL:          client.AccessTokenTTL,
+		CreationTime: accessTokenIssuedAt.Unix(),
+		EntityID:     entity.ID,
+
+		// TODO: Move this to InternalMeta once support is checked in
+		Meta: map[string]string{
+			accessTokenClientIDMeta: client.ClientID,
+			accessTokenScopesMeta:   strings.Join(authCodeEntry.scopes, scopesDelimiter),
+		},
+
+		// TODO: Add the InlinePolicy once support is checked in
+		//InlinePolicy: fmt.Sprintf(`
+		//	path "identity/oidc/provider/%s/userinfo" {
+		//		capabilities = ["read", "update"]
+		//	}
+		//`, name),
+	}
+	err = i.tokenStorer.TokenStore().create(ctx, accessToken)
```

Comment:
The more I stare at this the more I think that we should expose creation as `CreateToken` and thread that through core.

Reply:
I agree. I'll modify this to remove the `TokenStore()` getter and add `CreateToken()` like you've suggested.